src/test/descriptor_tests.cpp

```diff
@@ -40,32 +45,46 @@ bool EqualDescriptor(std::string a, std::string b)
     return a == b;
 }
 
-std::string MaybeUseHInsteadOfApostrophy(std::string ret)
+std::string UseHInsteadOfApostrophe(std::string ret)
 {
-    if (InsecureRandBool()) {
-        while (true) {
-            auto it = ret.find("'");
-            if (it != std::string::npos) {
-                ret[it] = 'h';
-                if (ret.size() > 9 && ret[ret.size() - 9] == '#') ret = ret.substr(0, ret.size() - 9); // Changing apostrophe to h breaks the checksum
-            } else {
-                break;
+    while (true) {
+        auto it = ret.find("'");
+        if (it != std::string::npos) {
+            ret[it] = 'h';
+            if (ret.size() > 9 && ret[ret.size() - 9] == '#') {
+                Span<const char> data = Span<const char>(ret.data(), ret.size() - 9);
+                std::string new_checksum = DescriptorChecksum(data);
+                ret = ret.substr(0, ret.size() - 8) + new_checksum; // Changing apostrophe to h breaks the checksum - recalculate it and replace the broken one
             }
+        } else {
+            break;
         }
     }
     return ret;
 }
 
 const std::set<std::vector<uint32_t>> ONLY_EMPTY{{}};
 
-void Check(const std::string& prv, const std::string& pub, int flags, const std::vector<std::vector<std::string>>& scripts, const std::set<std::vector<uint32_t>>& paths = ONLY_EMPTY)
+void DoCheck(const std::string& prv, const std::string& pub, int flags, const std::vector<std::vector<std::string>>& scripts, const std::set<std::vector<uint32_t>>& paths = ONLY_EMPTY,
+    bool replace_apostrophe_with_h=false)
 {
     FlatSigningProvider keys_priv, keys_pub;
     std::set<std::vector<uint32_t>> left_paths = paths;
 
+    std::unique_ptr<Descriptor> parse_priv;
+    std::unique_ptr<Descriptor> parse_pub;
     // Check that parsing succeeds.
-    auto parse_priv = Parse(MaybeUseHInsteadOfApostrophy(prv), keys_priv);
-    auto parse_pub = Parse(MaybeUseHInsteadOfApostrophy(pub), keys_pub);
+    if (replace_apostrophe_with_h)
```

Comment:
I think it would make more sense to largely leave this function untouched and instead do the `'` to `h` change in your new `Check` function below. The only change here would be to remove `MaybeUseHInsteadOfApostrophy`

Reply:
I agree and I did it that way initially but then, for example, this check failed:
```
std::string pub1 = parse_priv->ToString();
BOOST_CHECK(EqualDescriptor(pub, pub1));
```
since Descriptor::ToString returns a descriptor formatted with `'`, while pub has `h`, if I do it the way you suggest. I had to work around this by passing pub and prv without modification and introducing a boolean argument of whether to swap `'` with `h` only when passing them to the `Parse` function.